compiler-plugin/src/main/java/io/ballerina/lib/data/csvdata/compiler/CsvDataTypeValidator.java

```diff
@@ -0,0 +1,641 @@
+        Optional<Symbol> symbol = semanticModel.symbol(assignmentStatementNode.varRef());
+        if (symbol.isEmpty()) {
+            return;
+        }
+        TypeSymbol typeSymbol = ((VariableSymbol) symbol.get()).typeDescriptor();
+        validateFunctionParameterTypes(expressionNode, typeSymbol, currentLocation, ctx);
+    }
+
+    private void processLocalVarDeclNode(VariableDeclarationNode variableDeclarationNode,
+                                         SyntaxNodeAnalysisContext ctx) {
+        Optional<ExpressionNode> initializer = variableDeclarationNode.initializer();
+        if (initializer.isEmpty()) {
+            return;
+        }
+
+        currentLocation = variableDeclarationNode.typedBindingPattern().typeDescriptor().location();
+        Optional<Symbol> symbol = semanticModel.symbol(variableDeclarationNode.typedBindingPattern());
+        if (symbol.isEmpty()) {
+            return;
+        }
+
+        TypeSymbol typeSymbol = ((VariableSymbol) symbol.get()).typeDescriptor();
+        ExpressionNode expressionNode = initializer.get();
+        if (!isParseFunctionOfStringSource(expressionNode)) {
+            checkTypeAndDetectDuplicateFields(typeSymbol, ctx);
+            return;
+        }
+        validateExpectedType(typeSymbol, currentLocation, ctx);
+        validateFunctionParameterTypes(expressionNode, typeSymbol, expressionNode.location(), ctx);
+    }
+
+    private void checkTypeAndDetectDuplicateFields(TypeSymbol typeSymbol, SyntaxNodeAnalysisContext ctx) {
+        switch (typeSymbol.typeKind()) {
+            case RECORD -> detectDuplicateFields((RecordTypeSymbol) typeSymbol, ctx);
+            case ARRAY -> checkTypeAndDetectDuplicateFields(((ArrayTypeSymbol) typeSymbol).memberTypeDescriptor(), ctx);
+            case TUPLE -> {
+                for (TypeSymbol memberType : ((TupleTypeSymbol) typeSymbol).memberTypeDescriptors()) {
+                    checkTypeAndDetectDuplicateFields(memberType, ctx);
+                }
+            }
+            case UNION -> {
+                for (TypeSymbol memberType : ((UnionTypeSymbol) typeSymbol).memberTypeDescriptors()) {
+                    checkTypeAndDetectDuplicateFields(memberType, ctx);
+                }
+            }
+            case TYPE_REFERENCE -> checkTypeAndDetectDuplicateFields(
+                    ((TypeReferenceTypeSymbol) typeSymbol).typeDescriptor(), ctx);
+            case INTERSECTION -> checkTypeAndDetectDuplicateFields(getRawType(typeSymbol), ctx);
+        }
+    }
+
+    private FunctionCallExpressionNode getFunctionCallExpressionNodeIfPresent(ExpressionNode expressionNode) {
+        if (expressionNode.kind() == SyntaxKind.CHECK_EXPRESSION) {
+            expressionNode = ((CheckExpressionNode) expressionNode).expression();
+        }
+
+        if (expressionNode.kind() != SyntaxKind.FUNCTION_CALL) {
+            return null;
+        }
+        return ((FunctionCallExpressionNode) expressionNode);
```

Comment:
I think if you invert the last if we can turn this to a switch. WDYT?
```suggestion
       return switch(expressionNode.kind()) {
            case CHECK_EXPRESSION -> getFunctionCallExpressionNodeIfPresent((CheckExpressionNode) expressionNode.expression());
            case FUNCTION_CALL -> (FunctionCallExpressionNode)expressionNode;
            default -> null
       }
```